action/sealedenvelope.go

```diff
@@ -48,27 +52,43 @@ func (sealed *SealedEnvelope) envelopeHash() (hash.Hash256, error) {
 // Hash returns the hash value of SealedEnvelope.
 // an all-0 return value means the transaction is invalid
 func (sealed *SealedEnvelope) Hash() (hash.Hash256, error) {
+	if sealed.hash != hash.ZeroHash256 || sealed.hashErr != nil {
+		return sealed.hash, sealed.hashErr
+	}
 	switch sealed.encoding {
 	case iotextypes.Encoding_ETHEREUM_RLP:
 		act, ok := sealed.Action().(EthCompatibleAction)
 		if !ok {
-			return hash.ZeroHash256, ErrInvalidAct
+			sealed.hash, sealed.hashErr = hash.ZeroHash256, ErrInvalidAct
+			break
 		}
 		tx, err := act.ToEthTx()
 		if err != nil {
-			return hash.ZeroHash256, err
+			sealed.hash, sealed.hashErr = hash.ZeroHash256, err
+			break
 		}
-		return rlpSignedHash(tx, sealed.evmNetworkID, sealed.Signature())
+		sealed.evmNetworkID = config.EVMNetworkID()
+		sealed.hash, sealed.hashErr =
+			rlpSignedHash(tx, sealed.evmNetworkID, sealed.Signature())
 	case iotextypes.Encoding_IOTEX_PROTOBUF:
-		return hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto()))), nil
+		sealed.hash = hash.Hash256b(byteutil.Must(proto.Marshal(sealed.Proto())))
 	default:
-		return hash.ZeroHash256, errors.Errorf("unknown encoding type %v", sealed.encoding)
+		sealed.hashErr = ErrUnknownEncType(sealed.encoding)
 	}
+	return sealed.hash, sealed.hashErr
```

Comment:
as we discussed, please revert unnecessary changes, and make this an internal func `calcHash()`